### ISSUE TITLE
Markdown plugin should offer a way to generate code useful for iterating all markdown code

### ISSUE THREAD

**Issue by bitspittle** [varabyte/kobweb#375]
For a concrete description of the issue, on my blog site, I create a dynamic listing page based on all posts I've written. The code is a lot, inside my build script, which [you can see here](https://github.com/bitspittle/bitspittle.dev/blob/e5909f10b57aa52b2902ca5def7e27c6db7e482b/site/build.gradle.kts#L75).

Another user mentioned that they, too, found this unwieldly.

It might be nice to have a way to simplify this use-case, perhaps by automatically generating a top level variable of markdown paths to front matter values, for example, or maybe even by having the markdown plugin provide the parsing code, and the user just passes in a callback inside the build script.

I haven't thought this all the way through -- we probably want to think through more concrete examples -- but it would be nice if the build script code I linked to above could become something like:

```kotlin
kobweb {
   markdown {
      process { markdownData: MarkdownData ->
         // Put logic for generating the listing page in here?
      }
   }
}
```

where `MarkdownData` looks something like:

```kotlin
class MarkdownData(
   val path: String,
   val content: String,
   val frontMatter: Map<String, List<String>>,
)
```

We might provide a default implementation, so maybe a user could do something like:

```kotlin
kobweb {
   markdown {
      generateMarkdownListingFile("pkg.where.this.should.go")
   }
}
```

which would create a file that looks like:

```kotlin
package pkg.where.this.should.go

// Map of path of markdown file to frontmatter data
val MARKDOWN_DATA_LIST: List<MarkdownData> = ...
```

EDIT: Cleaned up the example a bit due to Ayfri's comment.

**Comment by Ayfri** [comment]
Shouldn't the `SITE_MARKDOWN_FRONTMATTER` variable be a Map instead of a single MarkdownData ?

**Comment by abd3lraouf** [comment]
shouldn't this be auto-generated using Ksp?

**Comment by bitspittle** [comment]
> shouldn't this be auto-generated using Ksp?

It could be. The original code was written before JS IR KSP was ready.

That said, while KSP is definitely a huge win for when parsing Kotlin code, I'm not as sure it matters as much when parsing markdown. And it will require handing down a huge amount of configuration from gradle to KSP. 

As such, there may not be any incentive to prioritize migrating it against other issues at the moment. (Unless someone points out a huge advantage I'm missing)

**Comment by bitspittle** [comment]
> Shouldn't the `SITE_MARKDOWN_FRONTMATTER` variable be a Map instead of a single MarkdownData ?

Sorry I never replied @Ayfri , yes you're right. The final version will be a map (at the very least, a collection)

**Comment by Ayfri** [comment]
Okay, no problem, thanks !

**Comment by bitspittle** [comment]
Fixed in 0.16.0
